### PR TITLE
Implement working resnet model

### PR DESCRIPTION
**Pull request by parthshah231**
I would like you review the code so far, particularly, [dataloader.py](https://github.com/parthshah231/chest_xray/compare/develop?expand=1#diff-44e6592c534d82d44cb8cae35c7d49bdfafd5bd072fb481b6265e075e8c7be88), [resnet.py](https://github.com/parthshah231/chest_xray/compare/develop?expand=1#diff-f0c59a5240882ba91c2abe5016bdd1ff0620e0b8ce5c46e8421d3962a5c4c278), [run_resnet.py](https://github.com/parthshah231/chest_xray/compare/develop?expand=1#diff-626eaad990c6ff5f9b89aef117c143ef62e9a91544dd8dcee595bb15fa4c7a4b)

### CODE REVIEW

File: callbacks.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        monitor="val/bce",
+        save_last=True,
+        save_top_k=1,
+        mode="min",
+        auto_insert_metric_name=True,
+        save_weights_only=False,
+        every_n_epochs=1,
+    )
+
+    callbacks_ = [
```
Comment (DM-Berger) [inline callbacks.py:29]: Any reason we can't just name this `callbacks`, without the underscore?

File: constants.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from pathlib import Path
+
+ROOT = Path(__file__).resolve().parent
+DATA = ROOT / "data"
+TRAIN_DIR = DATA / "train"
+VAL_DIR = DATA / "val"
+TEST_DIR = DATA / "test"
+
+# NO_VAL = True
+NO_VAL = False
```
Comment (DM-Berger) [inline constants.py:10]: This kind of thing would be better to have handled with an `ArgumentParser` option, or alternately via an environment variable. E.g.

```suggestion
NO_VAL = os.environ.get("NO_VAL") is not None
```

File: conv.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+from typing import Any, Optional, Tuple
+
+from pytorch_lightning import LightningModule
+from torch import Tensor
+
+
+class LitConv(LightningModule):
```
Comment (DM-Berger) [inline conv.py]: Obviously this class doesn't really do anything. If it was intended as an abstract base class to inherit from, you should perhaps make this clear with `abc` and `@abstractmethod` decorators.

```suggestion

from abc import ABC, abstractmethod

class LitConv(LightningModule, ABC):
# [...]
@abstractmethod
def forward(self, x):
    pass
```

If this is just scaffolding, then a `# TODO` comment or some `raise NotImplementedError()`s would also be good.

File: dataloader.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+from constants import NO_VAL, TEST_DIR, TRAIN_DIR, VAL_DIR
+
+# Make a hashmap for both datasets
+
+
+class ChestXrayDataset(Dataset):
+    def __init__(
+        self,
+        phase: str,
```
Comment (DM-Berger) [inline dataloader.py]: Since `phase` here can only actually be either `"train"` or `"val"`, it would be better to annotate this exactly, either with 

```suggestion
        phase: Literal["train", "val"],
```

or by defining a `Phase` enum, and using that:

```suggestion
from enum import Enum

class Phase(Enum):
    Train = "train"
    Val = "val"
    Test = "test"
    Predict = "pred"
    
class ChestXrayDataset(Dataset):
    def __init__(
        self,
        phase: Phase,
```

File: dataloader.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        if self.crop:
+            x = 0
+            y = 0
+            h, w = image.shape
+            h_6 = h // 6
+            w_6 = w // 6
+            cropped_image = image[y + h_6 :, x + w_6 : w - w_6]
+            image = cropped_image
+        image = torch.from_numpy(image).unsqueeze(0)
+        image = image.to(torch.float32)
```
Comment (DM-Berger) [inline dataloader.py]:
```suggestion
        image = torch.from_numpy(image).unsqueeze(0).float()
```

File: dataloader.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        label = self.labels[idx]
+        label = torch.FloatTensor([label])
+        resize_crop = RandomResizedCrop(size=(self.patch_size, self.patch_size))
```
Comment (DM-Berger) [inline dataloader.py]: Since this transform is constant for the class based on the patch size, we really don't want to be constructing a new one every `__getitem__` call. We should make this once in `__init__` only.

```suggestion
        # in __init__:
        self.resize_crop = RandomResizedCrop(size=(self.patch_size, self.patch_size))
        
        # in __getitem__
        resize_crop_img = self.resize_crop(image)
```

File: dataloader.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    def __init__(
+        self,
+        crop: bool = False,
+        n_per_image: int = 1,
+        patch_size: int = 64,
+        transfom: Callable = None,
+    ) -> None:
+        super().__init__()
+        if NO_VAL:
+            self.img_paths = list(TEST_DIR.rglob("*.jpeg"))[:10]
```
Comment (DM-Berger) [inline dataloader.py]: Suggestion: replace the magic number`[:10]` with a named constant, and document your reasoning here. E.g. why would we be changing the amount of testing depending on validation? If this is for fast development testing, maybe add a comment or, even better, implement a `--fast-dev-run` option via ArgumentParser.

File: dataloader.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        self,
+        phase: str,
+        crop: bool = False,
+        patch_size: int = 64,
+        transfom: Callable = None,
+    ) -> None:
+        super().__init__()
+        self.phase = phase
+        if self.phase == "train":
+            self.img_paths = list(TRAIN_DIR.rglob("*.jpeg"))
```
Comment (DM-Berger) [inline dataloader.py]: Can you be certain that the behaviour of `Path.rglob` is consistent? E.g. is is possible the ordering of paths returned is different each time, because `rglob` uses a hash of some kind internally? Remove doubt by making sure:

```suggestion
            self.img_paths = sorted(TRAIN_DIR.rglob("*.jpeg"))
```

File: dataloader.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        if len(image.shape) == 3 and image.shape[2] == 3:
+            image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+            image = image.squeeze()
+        if self.crop:
+            x = 0
+            y = 0
+            h, w = image.shape
+            h_6 = h // 6
+            w_6 = w // 6
+            cropped_image = image[y + h_6 :, x + w_6 : w - w_6]
```
Comment (DM-Berger) [inline dataloader.py]: Surely we can simplify this a lot:

```suggestion
            h, w = image.shape
            cropped_image = image[h // 6 :, w // 6 : w - w // 6]
```
or

```suggestion
            h, w = image.shape
            w6 = w // 6
            cropped_image = image[h // 6 :, w6 : w - w6]
```

if you are worried about the one extra op.

File: dataloader.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        self.n_per_image = n_per_image
+        self.transform = transfom
+
+    def __getitem__(self, idx: int) -> Tuple[Tensor, Tensor]:
+
+        # image = io.imread(self.img_paths[idx//2])
+        image = io.imread(self.img_paths[idx])
+        # look into different caching methods or
+        # do these processes beforehand and store in a location
+        # from where you can load it directly if the dataset
```
Comment (DM-Berger) [inline dataloader.py:112]: Good thinking. This is really hard to do well without losing randomness, or taking up a massive amount of memory.

File: mlp.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
+from typing import Any, Optional, Tuple
+
+import torch
+from pytorch_lightning import LightningModule
+from torch import Tensor
+from torch.nn import BCELoss, Linear, Sigmoid
+from torch.optim import AdamW
+
+
+class LitLinear(LightningModule):
```
Comment (DM-Berger) [inline mlp.py]: Nit: I don't like including "Lit" or anything that references Lightning in our class name, because our IDE and the inheritance is going to show us this already anyway. Ultimately we don't care about using Lightning, and if they change their name, we don't want our code to look irrelevant either. Better to just name what the model does than to reference a framework name:

```suggestion
class SingleLinear(LightningModule):  # or SimpleLinear, or OneLinear, etc.
```

File: mlp.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    def __init__(self, in_features: int, out_features: int) -> None:
+        super().__init__()
+        self.linear = Linear(in_features=in_features, out_features=out_features)
+        self.sigmoid = Sigmoid()
+        self.bce = BCELoss()
+
+    def forward(self, x):
+        x = self.linear(x)
+        out = self.sigmoid(x)
+        return out
```
Comment (DM-Berger) [inline mlp.py:20]: So torch has a really handy [`BCEWithLogitsLoss`](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) that combines the sigmoid and `BCELoss` into one operation that is more numerically-stable than the two individually. I would strongly recommend to use that.

```suggestion
        self.loss = BCEWithLogitsLoss()

    def forward(self, x):
        x = self.linear(x)
        out = self.loss(x)
        return out
```

File: mlp.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    def validation_step(self, batch: Tuple[Tensor, Tensor], *args, **kwargs) -> Tensor:
+        return self._shared_step(batch, "val")
+
+    def predict_step(self, batch: Tensor, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
+        # x -> List[Tensor] | x[0].shape -> [1, 1, 64, 64]
+        # target -> [1, 1]
+        x, target = batch
+        x = torch.stack(x)  # x -> [n, 1, 1, 64, 64] n -> number of patches
+        x = x.view(x.shape[0], -1)  # [n, 4096]
```
Comment (DM-Berger) [inline mlp.py:33]: The shape annotations here are really helpful and make it clear what you are doing. However, even clearer (and more idiomatic) would be to use [`Tensor.flatten`](https://pytorch.org/docs/stable/generated/torch.flatten.html#torch.flatten) or make a `self.flatten = Flatten()` layer, and use that. 

```suggestion
        x = torch.stack(x)  # x -> [n, 1, 1, 64, 64] n -> number of patches
        x = x.flatten(start_dim=1)  # [n, 4096]
```

File: mlp.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        preds = self(x)  # [n, 1]
+        return preds, target
+
+    def _shared_step(self, batch: Tuple[Tensor, Tensor], phase: str) -> Any:
+        # x      -> [16, 1, 64, 64]
+        # target -> [16, 1]
+        x, target = batch
+        x = x.view(x.shape[0], -1)  # [16, 4096]
```
Comment (DM-Berger) [inline mlp.py:41]: Same, `.flatten` is more idiomatic

File: mlp.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        preds = self(x)  # [16, 1]
+        loss = self.bce(preds, target)
```
Comment (DM-Berger) [inline mlp.py:43]: If using the combined `BCEWithLogitsLoss`, this would become:

```suggestion
        loss = self(x)
```

File: mlp.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        self.log(f"{phase}/bce", loss, prog_bar=True)
+        return loss
+
+    def configure_optimizers(self):
+        opt = AdamW(self.parameters())
```
Comment (DM-Berger) [inline mlp.py:48]: Usually you want to play around with the learning rate a bit. It is unusual not to see it passed in here.

File: resnet.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+from torch import Tensor
+from torch.nn import BCEWithLogitsLoss, Linear
+from torch.optim import AdamW
+from torch.optim.lr_scheduler import CosineAnnealingLR
+from torchio.transforms import RescaleIntensity
+
+from config import Config
+
+
+class LitResnet(LightningModule):
```
Comment (DM-Berger) [inline resnet.py]: Nit: Remove reference to Lightning in name
```suggestion
class Resnet(LightningModule):
```

File: resnet.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        self.resnet_version = resnet_version
+        self.resnet_model = resnets[resnet_version](pretrained=True)
+        linear_size = list(self.resnet_model.children())[-1].in_features
+        self.resnet_model.fc = Linear(linear_size, out_features)
+        self.config = config
+
+    def save_configs(self, log_dir: Path):
+        log_dir = Path(log_dir) / "config.json"
+        self.config.__dict__["resnet_version"] = self.resnet_version
+        json.dump(self.config.__dict__, open(log_dir, "w"))
```
Comment (DM-Berger) [inline resnet.py]: I am pretty sure you leak a file descriptor here, since you never `close` `log_dir`. Doesn't matter much here, but this is a bad thing in general. Just use a context manager to automate closing:
```suggestion
        with open(log_dir, "w") as handle:
            json.dump(self.config.__dict__, handle)
```

File: resnet.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        # DataFrame(self.config).to_json(log_dir_version)
+
+    def forward(self, x, *args, **kwargs) -> Any:
+        x = self.resnet_model(x)
+        return x
```
Comment (DM-Berger) [inline resnet.py:49]: I know typing kind of sucks for Lightning model core loop functions, but surely we can do better than this:
```suggestion
    def forward(self, x: Tensor, *args: Mapping, **kwargs: Mapping) -> Tensor:
        x = self.resnet_model(x)
        return x
```

File: resnet.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+    def training_step(self, batch: Tuple[Tensor, Tensor], *args, **kwargs) -> Tensor:
+        return self._shared_step(batch, "train")
+
+    def validation_step(self, batch: Tuple[Tensor, Tensor], *args, **kwargs) -> Optional[Tensor]:
```
Comment (DM-Berger) [inline resnet.py:54]: `self.shared_step` can never return `None`, so the `Optional` here seems incorrect?

File: resnet.py [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+        # for rescale_intersity it needs 3 dims (x, y, z)
+        # if the img is 2D we can add z -> 1 acc. to documentation
+        x = x.unsqueeze(3)
+        rescale = RescaleIntensity()
+        x_rescaled = rescale(x)
+        x = x_rescaled.squeeze(3)
+        # Since batch size is 1
+        x = x.unsqueeze(0)
+        preds = self(x)  # [n, 1]
+        preds = torch.sigmoid(preds) > 0.5
```
Comment (DM-Berger) [inline resnet.py:75]: So if we use `BCEWithLogitsLoss` then this code is broken. Instead we could make forward just return `self.linear(x)`, and have the best of both worlds.

File: resnet.py [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        # x      -> [16, 1, 64, 64]
+        # target -> [16, 1]
+        x, target = batch
+        # since the dataset contains grayscale images
+        # here the channel is 1 repeat_interleave() works
+        # as numpy.repeat() and expands the channels
+        # from 1 -> 3
+        x = torch.repeat_interleave(input=x, repeats=3, dim=1)
+        rescale = RescaleIntensity()
+        x_rescaled = rescale(x)
```
Comment (DM-Berger) [inline resnet.py]: MAJOR: This is either extremely clever or extremely questionable, because it rescales each patch *relative to itself*. 

- from a performance standpoint, this is a lot slower than just rescaling the intensity of the full original image, and then just grabbing patches from the intensity-rescaled image
- this rescales each patch. This means, for example, a pure air patch (e.g. all black) and a constant patch (e.g. all bright) end up looking mostly the same. But most likely we want different predictions for these
- this means different patches for the same subject can end up with completely different contrasts depending on the largest intensity in the patch. E.g. if patch1 and patch2 differ only by one pixel in the x direction (patch2 starts 1 pixel to the right of the start of patch1), but that last strip of pixels in patch2 has a single very bright pixel, then the entirety of patch2 will be "darkened" relative to that bright pixel (unless a robust intensity normalization method is used)

It is unclear if all of the above is good or bad. From a regularization standpoint, the intensity re-scaling per-patch might act like an augmentation, and be helpful. But if the data has a lot of spikes (not possible in .jpeg, but possible in actual raw MRI), then it means likely some patches end up being mostly black / mostly white (e.g. 0 or 1). Again, this may regularize, or destroy information, it isn't clear.

Basically, there should be a parameter that tests per-patch intensity normalization vs. per-subject, or at least a comment.

File: run_mlp.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+BATCH_SIZE = 64
+PATCH_SIZE = 64
+N_PATCHES = 5
+
+
+def run_MLP():
+    parser = ArgumentParser()
+    parser = Trainer.add_argparse_args(parser)
+    args = parser.parse_args(["--gpus=0", "--max_epochs=1", "--max_steps=5", "--val_check_interval=40"])
```
Comment (DM-Berger) [inline run_mlp.py:18]: I do this too all the time for testing, so :man_shrugging:, but arguably, committed code should not do this. Instead, there should be e.g. a `test_mlp.sh` file that looks something like

```bash
source .venv/bin/activate
python run_mlp.py --gpus=0 --max_epochs=1 --max_steps=5 --val_check_interval=40
```
or the hardcoded args should go in a `tests/test_mlp.py` to be run with e.g. `pytest` or `nose` or `unittest` or whatever.

File: run_mlp.py [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+    train_dataset = ChestXrayDataset(phase="train", crop=True, patch_size=PATCH_SIZE)
+    val_dataset = ChestXrayDataset(phase="val", crop=True, patch_size=PATCH_SIZE)
+    test_dataset = ChestXrayTestDataset(crop=True, patch_size=PATCH_SIZE, n_per_image=N_PATCHES)
+
+    train_dataloader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
+    val_dataloader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=True)
+    test_dataloader = DataLoader(test_dataset, batch_size=1)
```
Comment (DM-Berger) [inline run_mlp.py:25]: It would be much better (faster) if you can change the later logic so you don't need `batch_size=1` here.

File: run_mlp.py [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+    outputs = trainer.predict(model, test_dataloader)
+    # patch_preds
+    pred_patches, targets = zip(*outputs)
+    pred_patches = np.concatenate(pred_patches.numpy(), axis=0).astype(np.float32)
+    pred_patches = pred_patches.reshape(N_PATCHES, -1)
+    # [
+    #    [0, 0, 0, 1, 1],
+    #    [0, 0, 0, 1, 1],
+    #    [0, 0, 0, 1, 1],
+    # ]
```
Comment (DM-Berger) [inline run_mlp.py]: Very mysterious comment. Best not to commit something like this.

File: run_mlp.py [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+    preds = []
+    for pred_patch in pred_patches:
+        values, count = np.unique(pred_patch, return_counts=True)
+        idx = np.argmax(count)
+        pred = values[idx]
+        preds.append(pred)
```
Comment (DM-Berger) [inline run_mlp.py:41]: So I know we went over this and it is correct, but it seems like a comment mentioning that this is computing a mode would be helpful. There is also an obvious implementation that uses `array_split` to consider. 

The other reason to comment is that the mode does funny things in cases of ties (or in this case, what if counts are equal). Since you have only two classes, you can prevent this simply by having an ODD number of `N_PATCHES`, and it is worth considering that. In general when doing ensembling by modal vote, you might want to choose a prime number of ensembles to minimize the possible number of ties (or think about how to handle / log ties).

File: run_mlp.py [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+    count = 0
+    for pred, target in zip(preds, targets):
+        if pred == target:
+            count += 1
```
Comment (DM-Berger) [inline run_mlp.py:46]: Surely there is a faster vectorized NumPy solution here, like

```python
count = np.sum(preds == targets)
```

A for loop on tensors / NumPy arrays is a code smell for sure.

File: run_mlp.py [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+    accuracy = count / len(targets) * 100
+    print(f"Accuracy is: {accuracy}")
```
Comment (DM-Berger) [inline run_mlp.py:49]: Is there any reason we can't just use [`torchmetrics.function.accuracy`](https://torchmetrics.readthedocs.io/en/stable/classification/accuracy.html#functional-interface) here?

File: run_resnet.py [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+    i = 0
+    outputs = []
+    while i < N_PATCHES:
+        output = trainer.predict(
+            model,
+            test_dataloader,
+            ckpt_path="best" if not NO_VAL else None,
+        )
+        outputs.append(output)
+        i += 1
```
Comment (DM-Berger) [inline run_resnet.py]: I know, I know, for loops get boring. But from a readability standpoint, this really *should* be a for loop over `range(N_PATCHES)`. When an experienced developer sees a `while` loop, they except something more complicated.

File: run_resnet.py [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+    outputs_tensor = torch.Tensor(outputs)
+    outputs_tensor = outputs_tensor.view(-1, 2)
```
Comment (DM-Berger) [inline run_resnet.py:86]:
```suggestion
    outputs_tensor = torch.Tensor(outputs).view(-1, 2)
```

File: test_dataloader.py [hunk 31]
```diff
@@ -0,0 +1,11 @@
...
+import matplotlib.pyplot as plt
+import torchvision.transforms as transforms
+
+from dataloader import ChestXrayDataset, ChestXrayTestDataset
+
+if __name__ == "__main__":
+    # dataset = ChestXrayDataset(phase="train", crop=True, transfom=transforms.ToTensor())
+    # print(dataset[0])
+    test_dataset = ChestXrayTestDataset(crop=True, n_per_image=10)
+    print(test_dataset[0])
```
Comment (DM-Berger) [inline test_dataloader.py:11]: A test that requires a human to read outputs (i.e. is not automated) is not ideal. This should be in a `tests` folder, and runnable with something like `pytest` or `unittest`. It should also include some `assert`s that check some important functionality. For example:

```suggestion
def test_dataset() -> None:
    for patch_size in [16, 32, 46, 128]:
        dataset = ChestXrayTestDataset(crop=True, n_per_image=10, patch_size=patch_size)
        assert dataset[0][0].shape = (1, patch_size, patch_size)
        assert dataset[0][1].int64().item() in [0, 1]
```

**Review by DM-Berger — COMMENTED** [review]
Overall this is extremely clean and straightforward code. Almost all my comments are minor and unimportant, especially at this early phase. Variable names feel to me to be generally excellent and feel fluent / natural for a native English speaker. Obviously you've emulated / drawn a lot from my code style / conventions, so it is hard for me to complain about things here. 

But feel free to choose your own style! For example, another common academic DL approach is to read config from a `.yaml` or `.json` file instead of from CLI args. This has some advantages in that different configurations can be committed and saved to aid in reproducibility, and you don't have to type out CLI args or write helper scripts. 

The only "big thing" to consider is the patch-wise vs. subject-wise intensity normalization decision. It's actually a kind-of interesting idea to normalize each patch, and to think in general for patch-based training whether augmentation and normalization should be done at the image (subject) level, or at the patch-level (or both). This might be something worth exploring that is very unique, even publishable! After all, this could apply to all DL. E.g. for CIFAR/ImageNet, you normalize first, then augment (including cropping to patches). But what if you normalize again after augments? Probably it doesn't matter much, but it is still something to think about!

File: callbacks.py (hunk 1 above)
Comment (parthshah231) [reply callbacks.py:29]: None, should I change it to `callbacks` instead?

File: constants.py (hunk 2 above)
Comment (parthshah231) [reply constants.py:10]: I will make the changes here when I implement a working `ArgumentParser`

File: conv.py (hunk 3 above)
Comment (parthshah231) [reply conv.py]: Planning to implementing some code for it later, added a `#TODO` there

File: dataloader.py (hunk 4 above)
Comment (parthshah231) [reply dataloader.py]: Implemented `Enum`

File: dataloader.py (hunk 5 above)
Comment (parthshah231) [reply dataloader.py]: Done

File: dataloader.py (hunk 6 above)
Comment (parthshah231) [reply dataloader.py]: Done

File: dataloader.py (hunk 7 above)
Comment (parthshah231) [reply dataloader.py]: Makes sense, again will implement these will doing `ArgumentParser`

File: dataloader.py (hunk 8 above)
Comment (parthshah231) [reply dataloader.py]: Done

File: dataloader.py (hunk 9 above)
Comment (parthshah231) [reply dataloader.py]: Makes sense

File: dataloader.py (hunk 10 above)
Comment (parthshah231) [reply dataloader.py:112]: Thank you. Will do it in the end because it is of little to no significance as of now.

File: mlp.py (hunk 11 above)
Comment (parthshah231) [reply mlp.py]: Done

File: mlp.py (hunk 16 above)
Comment (parthshah231) [reply mlp.py:48]: Done

File: resnet.py (hunk 17 above)
Comment (parthshah231) [reply resnet.py]: Done

File: resnet.py (hunk 18 above)
Comment (parthshah231) [reply resnet.py]: Okay, done

File: resnet.py (hunk 19 above)
Comment (parthshah231) [reply resnet.py:49]: I thought I would do that at the end lol

File: resnet.py (hunk 20 above)
Comment (parthshah231) [reply resnet.py:54]: Yeah, makes sense

File: run_mlp.py (hunk 25 above)
Comment (parthshah231) [reply run_mlp.py]: Okay, yeah that was old logic forgot to remove that, my bad

File: run_mlp.py (hunk 26 above)
Comment (parthshah231) [reply run_mlp.py:41]: Okay

File: run_resnet.py (hunk 29 above)
Comment (parthshah231) [reply run_resnet.py]: Okay
